Team 및 Member CRUD 구현

a-white-bit opened the pull request:
https://devarium.atlassian.net/browse/BE-23

a-white-bit committed:
feat: 팀 도메인 및 엔티티 클래스 설정

a-white-bit committed:
feat: project - team 연관관계 1:1 설정

a-white-bit committed:
feat: team 에러 핸들링 설정

a-white-bit committed:
feat: team 도메인 DTO

a-white-bit committed:
feat: team 레포지토리 save 구현

a-white-bit committed:
feat: team 서비스 메서드 선언

a-white-bit committed:
feat: 팀 생성 서비스 로직

a-white-bit committed:
feat: 팀 삭제 서비스 로직

a-white-bit committed:
feat: 사용자가 팀 리더가 맞는지 확인

a-white-bit committed:
feat: getTeam 메서드 추가

a-white-bit committed:
feat: team 삭제 로직 추가(soft-delete)

a-white-bit committed:
feat: team 삭제 (hard-delete) 추가
- 스케줄링에 사용됨

a-white-bit committed:
feat: team 멤버 컬렉션 타입을 List -> Set으로 변경
- 중복 불가

a-white-bit committed:
feat: team 변경 로직 추가

a-white-bit committed:
chore: 의도하지 않은 코드 제거

a-white-bit committed:
fix: deletedAt 필드가 업데이트 되도록 수정 및 빌더 id 제거

a-white-bit committed:
feat: 사용자가 속한 팀 가져오기 (페이징)

a-white-bit committed:
feat: team 서비스 빈 등록

a-white-bit committed:
fix: getTeam(단일 팀 조회)은 TeamServiceImpl 내부에서만 사용되도록 수정

a-white-bit committed:
feat: team 에러코드 추가 및 변경
- FORBIDDEN_ACCESS: 사용자가 팀을 수정할 수 없을 때
- NOT_TEAM_MEMBERS: 요청 셋이 팀의 멤버가 아닐 때

a-white-bit committed:
chore: 공백 추가

a-white-bit committed:
feat: 입력한 userIds 셋이 팀의 멤버인지 확인하는 메서드 추가

a-white-bit committed:
refactor: 팀 리더 검증 메서드 예외처리를 TeamException 에서 처리하도록 수정

a-white-bit committed:
feat: 팀의 멤버인지 검증하는 로직 추가

a-white-bit committed:
feat: team RestController 작성

sillysillyman commented on api/src/main/java/io/devarium/api/common/GlobalExceptionHandler.java:
```diff
@@ -83,6 +85,19 @@ public ResponseEntity<ErrorResponse> handleProjectException(ProjectException e)
+                errorCode.getStatus().value(),
+                e.getMessage())
+            );
+
```
불필요한 공백 삭제

sillysillyman commented on api/src/main/java/io/devarium/api/controller/team/TeamController.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    @PostMapping("/{teamId}/members")
+    public ResponseEntity<SingleItemResponse<TeamResponse>> createMembers(
```
`create-`, `get-`, `update-`, `delete-` 등의 접두사는 도메인 모델 CRUD에 적절한 것 같습니다. `members`는 `Team`의 필드이니 다른 접두사를 붙이는 게 좋을 것 같습니다. 일단 이 API의 역할이 무엇인지 모호한 것 같습니다.

1. 사용자들을 멤버로 초대한다 -> `inviteMembers` -> 사용자들의 승인 여부에 따라 팀에 포함 로직 추가 필요
2. 팀 참여 요청을 보낸 사용자들을 멤버에 추가시킨다 -> `acceptMembers`

멤버 추가를 한번에 여러명 할 수 있게 한 것은 효율적인 것 같습니다. 추가로 일반 사용자가 팀에게 합류 요청하는 API도 필요할 것 같습니다.

a-white-bit replied:
메서드명이 모호한지는 동의할 수 없습니다만(충분히 알 수 있다고 생각합니다)
Member 도메인을 따로 빼는 작업 중이니 이 부분은 해소될 겁니다.

sillysillyman commented on api/src/main/java/io/devarium/api/controller/team/dto/TeamResponse.java:
```diff
@@ -0,0 +1,30 @@
+    String githubUrl,
+    Long leaderId,
+    Set<Long> memberIds,
+    Instant deletedAt
```
`deletedAt`을 포함한 이유는 삭제된 팀도 조회가 가능하도록 하는 건가요? 소프트 딜리드 된 팀을 조회하지  못 하도록 한다면 `deletedAt`을 응답에 포함하지 않아도 될 것 같습니다.

a-white-bit replied:
삭제된 리소스를 넘기면 프론트엔드에서 사용자 경험에 따라 사용할 수 있도록 책임을 주는 것도 방법이라 생각합니다.

지금은 통일성을 위해 맞추겠습니다.

a-white-bit replied:
추후 논의 하죠

sillysillyman commented on core/src/main/java/io/devarium/core/domain/team/Team.java:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    public void validateMembers(Set<Long> userIds) {
```
Set 크기가 그렇게 크지 않아서 성능 이슈는 없겠지만 코드 리뷰와 별개로 자바 스트림에 관한 좋은 글 추천드립니다.
https://sigridjin.medium.com/java-stream-api%EB%8A%94-%EC%99%9C-for-loop%EB%B3%B4%EB%8B%A4-%EB%8A%90%EB%A6%B4%EA%B9%8C-50dec4b9974b

성능 상 for loop를 끝낸 뒤에 `nonMemberIds`를 검사하는 것보다 foor loop 중에 early return 하는 것이 좋을 것 같습니다.

a-white-bit replied:
`Member` 도메인이 추가되어 이제 팀 멤버를 검증할 때 더 이상 스트림을 사용하지 않음

sillysillyman commented on core/src/main/java/io/devarium/core/domain/team/Team.java:
```diff
@@ -0,0 +1,85 @@
+
+    public void validateLeader(Long userId) {
+        if (!this.leaderId.equals(userId)) {
+            throw new TeamException(TeamErrorCode.FORBIDDEN_ACCESS, userId, this.id);
```
접근 권한에 대한 에러는 `AuthException`을 사용하는 게 좋을 것 같습니다.

a-white-bit replied:
같은 의미더라도 각 도메인에서 발생하는 세부적인 메시지를 커스텀하는 것이 디버깅에 좋을 것 같습니다.

sillysillyman commented on core/src/main/java/io/devarium/core/domain/team/Team.java:
```diff
@@ -0,0 +1,85 @@
+    public static final int DEFAULT_PAGE_SIZE = 1;
+
+    private final Long id;
+    private final Set<Long> memberIds;
```
`memberIds`가 외부에서 사용되지 않도록 명시적 Getter를 추가해주세요.

```java
public Set<Long> getMemberIds() {
    return Collections.unmodifiableSet(memberIds);
}
```

`unmodifiableSet`을 반환하여 불변성을 보장합니다.

sillysillyman commented on core/src/main/java/io/devarium/core/domain/team/Team.java:
```diff
@@ -0,0 +1,85 @@
+        this.picture = picture;
+        this.githubUrl = githubUrl;
+        this.leaderId = leaderId;
+        this.memberIds = memberIds;
```
Set은 mutable한 객체이기 때문에 외부의 변경으로부터 safe 하기 위해 방어적 복사가 필요합니다.

```java
memberIds != null ? new HashSet<>(memberIds) : new HashSet<>();
```

a-white-bit replied:
`Member` 도메인이 추가되어 더 이상 `Team`은 `memberIds` Set을 갖지 않음

sillysillyman commented on core/src/main/java/io/devarium/core/domain/team/exception/TeamErrorCode.java:
```diff
@@ -0,0 +1,22 @@
+@RequiredArgsConstructor
+public enum TeamErrorCode {
+    TEAM_NOT_FOUND(HttpStatus.NOT_FOUND, "Team not found with id: %d"),
+    FORBIDDEN_ACCESS(HttpStatus.FORBIDDEN,
```
`AuthException` 사용은 어떨까 싶습니다.

sillysillyman commented on core/src/main/java/io/devarium/core/domain/team/Team.java:
```diff
@@ -0,0 +1,85 @@
+@Getter
+public class Team {
+
+    public static final int DEFAULT_PAGE_SIZE = 1;
```
페이지 사이즈가 1일 때의 페이징 조회가 어떤 화면으로 구성될지 생각해두신 게 있으실까요?

a-white-bit replied:
`Member` 도메인이 추가되어 `Team`의 페이징 처리 API를 제거한 상태입니다.
여러 팀 정보를 한 번에 조회할 일이 있을지 지금 떠오르진 않네요.
필요한 API가 생기면 다시 추가하겠습니다.
`Team.DEFAULT_PAGE_SIZE`는 삭제됨.

a-white-bit replied:
대신 단일 팀 조회 API 추가

sillysillyman commented on infrastructure/src/main/java/io/devarium/infrastructure/persistence/entity/ProjectEntity.java:
```diff
@@ -33,7 +35,9 @@ public class ProjectEntity extends BaseEntity {
-    // TODO: 1:1 mapping to Team
+    @OneToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "team_id", nullable = false, unique = true)
+    private TeamEntity team;
```
추후에 한 팀이 여러 프로젝트를 하는 것도 고려해볼 필요가 있을 것 같습니다. 예를 들어 한 팀이 프로젝트를 종료하고 새로운 팀 생성 없이 그 팀 그대로 다른 프로젝트를 진행할 수도 있으니까요.

a-white-bit replied:
제 생각에는 '사이드 프로젝트'의 성격 상 팀의 수명이 짧을 것 같다고 생각하여
흔하지는 않을 것 같아 1:1로 설계해두었습니다.

sillysillyman commented on infrastructure/src/main/java/io/devarium/infrastructure/persistence/entity/TeamEntity.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    // TODO : 상위 클래스에 두고 모든 엔티티가 일관적으로 동작하도록 수정
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        TeamEntity team = (TeamEntity) o;
+        return id != null && id.equals(team.id);
+    }
+
+    @Override
+    public int hashCode() {
+        return id != null ? id.hashCode() : 0;
+    }
```
명시적으로 구현하는 것이 좋은 방법인지는 모르겠습니다. 슬랙에 말씀해주셨던 이유를 제가 제대로 이해하지 못 해서 그런데 구현하신 이유를 자세하게 설명해주시고 저도 같이 고민해보고 싶습니다.

a-white-bit replied:
`Member` 도메인이 추가되어 `Team`의 `memberIds` set을 다룰 필요가 없어져서 삭제되었습니다.

a-white-bit committed:
feat: Member 도메인 및 엔티티 추가
- user - team 중간테이블 구성
- 팀 내 역할 부여

a-white-bit committed:
feat: Member 도메인 및 엔티티 추가
- user - team 중간테이블 구성
- 팀 내 역할 부여

a-white-bit committed:
feat: member 도메인 에러 추가

a-white-bit committed:
feat: member 레포지토리 구현

a-white-bit committed:
feat: member DTO 구현

a-white-bit committed:
feat: member 서비스 로직 구현

a-white-bit committed:
feat: member 컨트롤러 구현

a-white-bit committed:
fix: team 사이드에서 user와 연관된 로직은 전부 member 사이드로 이전
- createMember, deleteMember
- team 도메인 및 엔티티에서 members 필드 삭제
- getTeams(사용자의 속한 팀 전체 조회) API를 getMembersByUser(사용자가 가입한 팀 멤버십 전체 조회) API로 변경

a-white-bit committed:
feat: getTeam(팀 단일조회) API 추가

a-white-bit committed:
fix: saveAll, deleteAll jpa 쿼리메서드를 기본 메서드로 사용하도록 수정

a-white-bit committed:
fix: member 생성, 수정, 삭제 API가 더 이상 페이징 전체 조회를 리턴하지 않도록 변경

a-white-bit committed:
refactor: 팀 소속 여부 메서드명 변경 (validateMember -> validateMembership)

a-white-bit committed:
Merge branch 'dev' of https://github.com/devarium/devarium-backend into BE-23

a-white-bit committed:
fix: delete http 요청 메서드가 put으로 잘못 표기된 점 수정

a-white-bit committed:
feat: dto 유효성 검증 추가

a-white-bit committed:
feat: 멤버 역할에 레벨을 부여하여 계층적으로 표현

a-white-bit committed:
chore: 사용하지 않는 에러코드 삭제

a-white-bit committed:
fix: 멤버 권한을 요구 권한 이상은 모두 통과되도록 수정

a-white-bit committed:
chore: 가독성 위한 공백 추가

a-white-bit committed:
feat: 팀 생성 요청자가 첫 멤버(리더)로 등록

a-white-bit committed:
fix: 팀 및 멤버 생성이 하나의 트랜잭션에서 실행되도록 수정

a-white-bit committed:
feat: 팀 리더를 변경할 때 변경에 따른 멤버 역할을 함께 변경

a-white-bit committed:
refactor: UpdateLeaderRequest DTO가 leaderId 값만 받도록 유효성 검사 추가
- memberIdToRole에 이상한 값이 들어오더라도 무시하고 컬렉션을 비워주는 로직 -> 이상한 값을 받지 않도록 검사

a-white-bit committed:
fix: 멤버 추가 및 수정 시 의도치 않은 데이터 저장 방지
- updateMembers는 멤버 역할을 리더로 바꾸거나 나의 권한 이상인 자를 수정할 수 없도록 변경
- createMembers는 리더를 추가할 수 없도록 변경
- isLeader 필드 추가

a-white-bit committed:
fix: 팀 리더 변경 시 MemberService.updateMember 대신 MebmerServie.updateLeader을 사용

a-white-bit committed:
chore: 필요없는 유효성 검증 파일 삭제

a-white-bit committed:
feat: 멤버 API 요청 시 연관 user 및 team이 삭제되었는지 확인

a-white-bit commented:
현재 User 및 Team 이 논리 삭제되었는지를 판단하는 로직이 추가되었지만 좋은 방법은 아닙니다.
쿼리 최적화를 할 때 수정해보겠습니다.

sillysillyman commented on api/src/main/java/io/devarium/api/common/PageTypeConverter.java:
따로 구현하신 이유가 있을까요? `MemberController`에서만 쓰이는 것 같은데 아래 매핑 함수를 이용하는 것이 더 간단해 보입니다.

```java
Page<MemberResponse> response = members.map(MemberResponse::from);
```

sillysillyman commented on api/src/main/java/io/devarium/api/controller/member/MemberController.java:
```diff
@@ -0,0 +1,99 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RequestMapping("/api/${api.version}")
```
`@PathVariable`로 `{teamId}`를 받기 때문에 URI에 리소스를 표현하는 것이 RESTful 할 것 같습니다. 또한, 컨트롤러 내의 모든 URI에 `members`를 공통으로 가지고 있습니다.

```java
@RequestMapping("/api/${api.version}/teams/{teamId}/members")
```

a-white-bit replied:
RequestMapping에서도 변수를 사용할 수 있는지는 생각 못했네요. 좋습니다.

그런데 다음 이유로 /teams 만 추가하도록 할게요.

sillysillyman commented on api/src/main/java/io/devarium/api/controller/member/MemberController.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    @GetMapping("/members")
+    public ResponseEntity<PagedListResponse<MemberResponse>> getMembersByUser(
```
사용자가 어느 팀의 멤버로 있는지 조회하는 API인 것 같은데 `TeamController`에서 **사용자가 속하는 팀 목록 조회 API**로 다루는 것이 좋을 것 같습니다. `MemberController`의 URI도 `@RequestMapping`으로 일치시킬 수 있고요.

a-white-bit replied:
하나의 유저는 하나의 팀에 여러 번 가입할 수 없습니다. 따라서 멤버 조회가 곧 팀 조회입니다.

`getMembersByUser()`는 사용자가 '내가 속한 팀' 페이지를 조회할 때 먼저 '내가 가입한 팀 목록', 즉 '나의 멤버십 조회'로써 어떤 팀에 가입했고 역할이 무엇인지 정도에 대한 간략한 목록이 나오는 것이고
그 목록에서 특정 멤버십을 클릭하면 '해당 팀 조회'를 하는 것이 좋을 것 같습니다.

그리고 기술적으로,
사용자가 속하는 팀 목록 조회 `getTeams()`를 사용하려면 이녀석은 JPQL 등 상세 쿼리 작성기술을 써야 합니다.

`getTeams()`를 삭제하고 Member로 넘긴 이유는 사용자가 속하는 팀 목록을 페이징 처리할 때 정렬을 '사용자가 가입한 순서' 또는 '프로젝트 수정된 순서' 등, 적어도 팀의 생성/수정/삭제 시간의 정렬은 어색하다는 점 때문입니다.
팀과 멤버의 관계는 멤버가 관계의 주인이며 단방향입니다.
위 페이징 정렬 순서를 해소하려면 JpaRepository에서 쿼리를 날릴 때, JPQL 등을 사용하여 상세 쿼리를 사용해야합니다.
페이징 뿐만 아니라 데이터를 조회하는 쿼리에서도 동일한 문제입니다.

상세쿼리를 사용하는 것은 쿼리 최적화도 될 수 있고 문제가 되지는 않으나,
Member 사이드에서 기본 Jpa를 사용하여 처리할 수 있는 로직을 굳이 관계를 표현하기 힘든 Team 사이드로 끌어서 사용하기에는 억지스러운 면이 있다고 생각합니다.

a-white-bit replied:
이미 예전에 작성했던 것이라 `getTeams()`도 살려 놓겠습니다.
편의에 따라 사용하도록 하죠

sillysillyman commented on api/src/main/java/io/devarium/api/controller/member/MemberController.java:
컨트롤러에서 생성, 조회, 수정은 프론트엔드에서 결과를 바로 조회에 사용할 수 있도록 리턴 타입을 변경해주세요. 생성 API 호출 후 `ResponseEntity<Void>`를 반환하면 프론트엔드에서는 조회 API를 한 번 더 호출해야 한다는 네트워크 오버헤드가 발생합니다. 따라서 생성/수정 API에서도 해당 작업이 완료된 후의 최신 데이터를 바로 응답값으로 반환하는 것이 좋습니다.

a-white-bit replied:
제 생각에는 작업이 완료된 후에는 요청한 팀에 속한 "모든 멤버"를 보여주는 것이 좋다고 생각합니다.
따라서 응답값으로는 적어도 "성공했다"는 정도만 알려주고, 클라이언트 혹은 프론트엔드에서 팀 멤버 조회 API를 날리는 것이 더 자연스러운 것 같습니다.

네트워크 오버헤드 단점이 있다는 것은 알고 있고, 중간 커밋에도 매 CRUD마다 전체 멤버 페이지네이션 리턴 작성한 기록이 있는데 역할이 분리되어있지 않고 중복 동작한다는 느낌이 강합니다.

조회 API는 어차피 자주 일어날 수 밖에 없고 수정 API 요청 후 조회 API 요청은 큰 부담은 아니라고 생각합니다.

sillysillyman commented on api/src/main/java/io/devarium/api/controller/member/dto/CreateMembersRequest.java:
`userIdToRole` 내부의 유효성 검사를 할 수 없습니다. 아래 구조는 어떠신가요? 

```java
public record CreateMembersRequest(
    @NotNull(message = "members must not be null")
    List<CreateMemberRequest> members
) implements CreateMembers {

    public record CreateMemberRequest(
        @NotNull(message = "userId must not be null")
        Long userId,
        
        @NotNull(message = "role must not be null")
        MemberRole role
    ) {
    }
}
```

ref. https://github.com/devarium/devarium-backend/pull/12/files#diff-65eb175e7eac977660ab23e886ea162e39c2ea1f0c4fbad244497adc3d7e9667

sillysillyman commented on api/src/main/java/io/devarium/api/controller/team/TeamController.java:
```diff
@@ -0,0 +1,86 @@
+        @AuthenticationPrincipal EmailPrincipal emailPrincipal
+    ) {
+        User user = emailPrincipal.getUser();
+        Team team = teamService.createTeam(request, user);
```
이 구간도 코드 내의 동일 부분처럼 바꾸시는 게 일관되어 보입니다.

```java
 Team team = teamService.createTeam(request, emailPrincipal.getUser());
```

sillysillyman commented on api/src/main/java/io/devarium/api/controller/team/dto/UpsertTeamRequest.java:
`name`은 유효성 검사가 필요할 것 같습니다.

a-white-bit replied:
Create + Update 유효성 검사가 달라 놔둔 것이었습니다.
분리하여 검사 추가할게요~

sillysillyman commented on core/src/main/java/io/devarium/core/domain/member/Member.java:
```diff
@@ -0,0 +1,61 @@
+    private MemberRole role;
+
+    @Getter(AccessLevel.NONE)
+    private Boolean isLeader;
```
```java
private Boolean leader;
```
로 선언하시면 `@Getter`로 인해 `isLeader()` 메서드 구현 없이 사용할 수 있습니다.

a-white-bit replied:
getIsLeader가 어색해서 추가하였습니다.

sillysillyman commented on core/src/main/java/io/devarium/core/domain/member/service/MemberServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    @Override
+    public void createFirstMember(Long teamId, Long userId) {
```
이 메서드가 처음 팀을 생성할 때만 쓰인다면 개인적으로 `createLeader`라는 메서드명이 더 좋을 것 같습니다.

sillysillyman commented on core/src/main/java/io/devarium/core/domain/member/service/MemberServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+
+    @Override
+    public void createFirstMember(Long teamId, Long userId) {
+        if (memberRepository.countByTeamId(teamId) != 0) {
```
`countBy`는 개수를 전부 세야 해서 `existBy`보다 느립니다.

sillysillyman commented on infrastructure/src/main/java/io/devarium/infrastructure/persistence/entity/MemberEntity.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Column(nullable = false)
+    @Getter(AccessLevel.NONE)
+    private Boolean isLeader;
```
컬럼의 대부분이 `null`일 텐데 이 필드가 필요한지 의문입니다. 팀의 리더는 Team 도메인에서 관리하고 검증하는 것은 어떨까요?

a-white-bit replied:
isLeader가 없을 때 개발상 불편한 부분이 생기고 쿼리가 복잡해지는 것을 편의적으로 필드 하나 추가하는 것으로 해결할 수 있어 종합적으로는 이 방법이 좀 더 나은 것 같습니다.
Member에서 역할 등을 관리하는데 리더 여부를 자주 확인하고 있고, 역할 변경으로 리더가 변경됨으로써 관계에 대한 처리가 필요합니다.

sillysillyman reviewed (commented):
일단 전반적인 리뷰 남깁니다.

a-white-bit committed:
refactor: 유틸 클래스를 사용하지 않고 간단하게 매핑하는 것으로 변경

a-white-bit committed:
feat: 사용자의 참여 팀 목록 조회 getTeams

a-white-bit committed:
fix: 멤버 API 엔드포인트 경로 변경

a-white-bit committed:
fix: createMember 요청 DTO 구조 변경

a-white-bit committed:
fix: createMember 요청 DTO 구조 변경 및 인터페이스 분리

a-white-bit committed:
feat: 멤버 삭제 요청 DTO 유효성 검사 추가

a-white-bit committed:
fix: UpdateMember 요청 DTO 유효성 검사 추가 및 타입 구조 변경

a-white-bit committed:
chore: 공백 삭제

a-white-bit committed:
feat: CreateTeam 요청 DTO 유효성 검사 추가

a-white-bit committed:
refactor: 메서드명 변경 createFirstMember -> createLeader

a-white-bit committed:
refactor: count 쿼리를 exists로 변경

a-white-bit committed:
refactor: 의미 없는 변수 선언을 생략

a-white-bit committed:
chore: dev merge to BE-23

a-white-bit committed:
fix: 필드명 변경으로 인한 오류 수정
